src/backend/access/transam/xlogreader.c

```diff
@@ -905,6 +905,9 @@ XLogDecodeNextRecord(XLogReaderState *state, bool nonblocking)
 			SKIP_INVALID_RECORD(RecPtr);
 		}
 
+		if (state->force_record_reassemble)
+			memcpy(state->readRecordBuf, record, total_len);
+
```

Comment:
I don't think this is needed. XLogReadRecord() returns a pointer to a buffer that holds the record, which can point to `readRecordBuf` if the record was reassembled, or straight to the page buffer otherwise. In any case, the pointer should be valid.

Reply:
Unfortunately it is needed. In the second case  XLogReadRecord returns pointer not to the page buffer, but for the buffer used to records unpacking. So it contains not the raw data, but unpacked data. 

Certainly my first attempt was to use pointer returned by `XLogReadRecord()`. But it didn't work:(

Reply:
Ok, I see